### ISSUE TITLE
Testing on ubuntu

### ISSUE THREAD

**Issue by albfan** [albfan/miraclecast#8]
Now ubuntu vivid vervet comes with systemd by default

https://wiki.ubuntu.com/SystemdForUpstartUsers

Some compilation errors was reported.

Let's give it a try

**Comment by albfan** [comment]
Seems it's not a definitive step (yet).

http://www.webupd8.org/2015/03/ubuntu-1504-vivid-vervet-switches-to.html?m=1

Fingers crossed!

**Comment by ljbhander** [comment]
The  ubuntu vivid vervet  seems not have all the heards in /usr/include/systemd.
I got the systemd lib(219) from :
git://anongit.freedesktop.org/systemd/system
Compile the systemd:
./autogen.sh
./configure CFLAGS='-g -O0 -ftrapv' --enable-compat-libs --enable-kdbus --sysconfdir=/etc --localstatedir=/var --libdir=/usr/lib --enable-gtk-doc --with-rootprefix= --with-rootlibdir=/lib

make install and reboot the system, unfortunately the system can't start up .

Then i reinstall the system . download the systemd lib and make install .
Get the miracle code from:
git://people.freedesktop.org/~dvdhrm/miracle

Compile the miracle code follow with https://github.com/albfan/miraclecast .
Got some file :miracled、miracle-dhcp、miracle-sinkctl、miracle-wifictl、miracle-wifid.
Steps:
wpa_supplicant 
sudo miracle-wifid &
sudo miracle-sinkctl
here not got the “[ADD]  Link: 3” but got [miracled]

Is this the environment problem?

**Comment by albfan** [comment]
You say you install systemd with --enable-kdbus and OS wont start. Then you install it again and then it works. Don't know the magic between to steps, but here is the standard method to compile packages from source on ubuntu

    $ mkdir systemd-src
    $ cd systemd-src
    $ apt-get source systemd
    $ sudo apt-get build-dep systemd
    $ DEB_BUILD_OPTIONS="--enable-kdbus" dpkg-buildpackage -rfakeroot -uc -b -j8
    $ sudo dpkg -i systemd-<whatever-version>.deb
    
A good detailed response about this http://askubuntu.com/a/28373/89479

Once you succesfully install systemd with kdbus I suggest to get miraclecast from this repo. There's some tweaks and fixes from dvdhrm repo like 12dbe6e335c67cd621c985471326f67e5920cf7f that can be annoying or difficult to solve.

If you can config wpa to be in "miraclecast mode" then you will see "[ADD] link <whatever>" if there is any miracast device searching for pairing

I just upgraded to ubuntu 15.04 (I'm writing from it). I will let you now my progress as soon as possible.

**Comment by ljbhander** [comment]
Thanks for your guidance.
When  i download systemd in my way, after “make install” , if i reboot the system, then it will not start up again.
So the second time after reinstalling  ubuntu vivid vervet, i would't reboot the system, and compile the maricle with command “./configure --prefix=/usr/local CFLAGS=-I/usr/local/include LDFLAGS=-L/usr/local/lib”.
With some steps, it will compile successful.

Next i will reinstall the system and systemd lib by your way.

 config wpa to be in "miraclecast mode" : how to config it?

**Comment by albfan** [comment]
See inside res directory. There is a bunch of resources you need to setup.

- Copy the dbus policy org.freedesktop.miracle.conf to /etc/dbus-1/system.d/
- Check your card or dongle has AP capabilities:
http://unix.stackexchange.com/questions/92727/how-can-i-tell-if-my-wifi-dongle-has-access-point-capabilities

- Kill actual wpa_supplicant process

- Config wpa_supplicant with wpa.conf file from res/ directory

There is a miracle-wifi.sh to config wpa on arch linux, changing interface
name for ubuntu (say wlan0 instead of wlp3s0) must work.

**Comment by ljbhander** [comment]
Got some mistake when "DEB_BUILD_OPTIONS="--enable-kdbus" dpkg-buildpackage -rfakeroot -uc -b"：
**************************************************************************************
==== test-verbs.log ====
Unknown operation command-not-found.
Too few arguments.
Too many arguments.
Requires operation parameter.
==== test-xml.log ====
debian/rules:250: recipe for target 'override_dh_auto_test' failed
make[1]: *** [override_dh_auto_test] Error 1
make[1]: Leaving directory '/home/nbpt/Downloads/systemd-src/systemd-219'
debian/rules:261: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
**************************************************************************************

**Comment by albfan** [comment]
Me too. I have not solved it yet.

It's preferred to use this source because it has a lot of patches (see
debian/patches directory there are a lot there.) Ubuntu is known to tweak a
lot its packages.

Meanwhile I guess you are in the same situation as I am. Your systemd
compilation is not recognized. You should use

    ./configure --prefix=/usr --build=x86_64-linux-gnu

To compile systemd

Because pkg-config on ubuntu (64 bits) looks on /usr/lib/x86_64-linux-gnu
for installed packages

You can test it issuing

    $ pkg-config pkg-config --variable pc_path

http://askubuntu.com/questions/210210/pkg-config-path-environment-variable

The output is your search path for installed libraries, and compiled
systemd is not there.

With that should be possible to compile miraclecast on ubuntu

**Comment by thopiekar** [comment]
Hey, I'm following your repo here in github and here is a hint from me:

Except of using ```"DEB_BUILD_OPTIONS="--enable-kdbus"``` to expand the list of configure options open the debian/rules file and add the option at the right position :smile:

**Comment by albfan** [comment]
A quick try of

    --build=x86_64-linux-gnu

from freedesktop repo did not success. Seems it do not respect --build option on 

    $ make install

It's time to face the big problem. Compare tweaks  on ubuntu fork to honor build option or ask systemd ubuntu maintainers about our problems to build the package With dbus support.

Former seems easier but latter would be most acuratte.

I will check if problems come from adding `--enable-kdbus` prior to ask anything

**Comment by thopiekar** [comment]
@albfan Just took a look on it. You'll need to add ```--enable-kdbus``` here:
```
CONFFLAGS_deb = \
	--enable-kdbus \
	--enable-selinux \
	--enable-introspection \
	--enable-gtk-doc \
	PYTHON=python3 \
	$(NULL)
```
```CONFFLAGS_udeb``` is for udebs (used for Installer CD's/USB sticks). However, I think builds will fails, because ```--fail-missing``` of this line:
```
override_dh_install:
[...]
	dh_install $(BOOTSTRAP_DH_FLAGS) -Nudev-udeb -Nlibudev1-udeb --sourcedir=debian/install/deb --fail-missing
```
So, new files have to be added to the needed *.install file in the debian folder.

I'm making a local build at the moment, if it does not take much time I'll make all changes and add a detailed instruction how to make it on your own, e.g. for newer upstream (== Ubuntu) releases.

**Comment by thopiekar** [comment]
How are the needed headers called? Just to verify my packages.

**Comment by thopiekar** [comment]
@ljbhander You might be interested in adding the ```nocheck``` option, see http://man.he.net/man1/dh_auto_test

**Comment by albfan** [comment]
Good to know. If it keeps failing on test suite we can try that, but for the process 1 on OS would the better to solve errors on test suite (advise for the brave ones)

**Comment by thopiekar** [comment]
![systemd-sd](https://cloud.githubusercontent.com/assets/1847437/6736185/3e2dcba4-ce5b-11e4-95df-2759ded48e06.png)

Ok, got the packages built. Had to do more then the instructions above :smiling_imp: 
https://mega.co.nz/#F!N1AGFIII!AVnoRGUysWfbspi2ZjQkDw

PS: amd64 packages are on the go ;)

**Comment by albfan** [comment]
I will fork the systemd project to keep a known source with necessary changes for ubuntu (and maybe others).

@thopiekar you can reproduce your steps there (or create the fork yourself) I want to stay as close to source as possible. There, releases can be created and shared.

To contribute this project, extend README.md with instructions for ubuntu linking the systemd fork.

**Comment by thopiekar** [comment]
Ok, by any reason my shared link to MEGA got removed. It includes a patch which shows, what needs to be done to make a kdbus rebuild and rebuilt packages are also there.
Maybe GitHub is removing comments with shared links automaticly..

**Comment by thopiekar** [comment]
However, as you wanted to have a vanilla/unpatched build of systemd, I invested some time building one via launchpad.net using a recipe. I took imported source code from freedesktop and used my modified packaging files without debian or ubuntu patches.

Here it is: https://code.launchpad.net/~thopiekar/+recipe/systemd-daily

**Comment by albfan** [comment]
I do remove it @thopiekar. Much better this way. See if you use:

    $ apt-get source systemd

There is a dsc file where you can see the git repo used on ubuntu (tag v219-4ubuntu5 I think)

Yesterday I try to understand ubuntu way of release packages but is rwally overcomplicated

- They have a repo created from imported tar.gz from what hey call upstream
- Then they tweak it (a lot i must say)
- Then they create a debian package with... more patches?
- All this stuff is released in some way with apt-get

If can start from one of this points and apply your fixes the code will be exactly the same that ubuntu releases but with dbus enabled

**Comment by thopiekar** [comment]
Yes, it is complicated. The reason is that it often takes much time until
patches are applied on the real upstream.
It is almost the same as Debian and Ubuntu do with the Linux kernel. In
Debian wheezy they use 3.2, but there are changes included which you see on
3.3 releases.
However, which systemd codebase do you really want to used now? The
freedesktop's or Ubuntu's?
Am 20.03.2015 09:17 schrieb "Alberto Fanjul" <notifications@github.com>:

**Comment by albfan** [comment]
I'm mainly arch linux user and freedesktop works for me as it is, but this issue is for ubuntu testing so I think is better to stay close to actual codebase for systemd package on ubuntu.

**Comment by albfan** [comment]
Well all is clear, for arch linux there is a separate issue #6. This one should use freedesktop source.

For ubuntu users best aproach is to provide a ppa like @thopiekar's one. It uses bazar instead of git, but it's the ubuntu way.

I think a better name for ppa is systemd-with-dbus

**Comment by thopiekar** [comment]
To be correct the bzr repo I use for the builds is just a mirror.
Launchpads build system can just handle bzr branches and therefore the code
from freedesktops git needs to be imported first ( which is done by
launchpad itself ).
Am 20.03.2015 18:44 schrieb "Alberto Fanjul" <notifications@github.com>:

**Comment by thopiekar** [comment]
```
thopiekar@32bit:~/miraclecast$ make 
[  4%] Building C object src/shared/CMakeFiles/miracle-shared.dir/rtsp.c.o
In file included from /home/thopiekar/miraclecast/src/shared/rtsp.c:31:0:
/usr/include/systemd/sd-event.h:73:67: error: unknown type name ‘siginfo_t’
 typedef int (*sd_event_child_handler_t)(sd_event_source *s, const siginfo_t *si, void *userdata);
                                                                   ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_new_unknown’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:389:2: warning: implicit declaration of function ‘strdup’ [-Wimplicit-function-declaration]
  m->unknown_head = strdup(head);
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:389:18: warning: assignment makes pointer from integer without a cast
  m->unknown_head = strdup(head);
                  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_new_request_n’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:419:2: warning: implicit declaration of function ‘strndup’ [-Wimplicit-function-declaration]
  m->request_method = strndup(method, methodlen);
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:419:20: warning: assignment makes pointer from integer without a cast
  m->request_method = strndup(method, methodlen);
                    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:423:17: warning: assignment makes pointer from integer without a cast
  m->request_uri = strndup(uri, urilen);
                 ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_new_raw_reply’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:469:19: warning: assignment makes pointer from integer without a cast
   m->reply_phrase = strdup(get_code_description(code));
                   ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:471:19: warning: assignment makes pointer from integer without a cast
   m->reply_phrase = strdup(phrase);
                   ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_is_request’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:612:9: warning: implicit declaration of function ‘strcasecmp’ [-Wimplicit-function-declaration]
         (!method || !strcasecmp(m->request_method, method)) &&
         ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_header_set_value’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:743:11: warning: assignment makes pointer from integer without a cast
  h->value = strndup(value, valuelen);
           ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_append_header’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:791:9: warning: assignment makes pointer from integer without a cast
  h->key = strndup(key, keylen);
         ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_append_header_line’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:837:2: warning: implicit declaration of function ‘strchrnul’ [-Wimplicit-function-declaration]
  value = strchrnul(line, ':');
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:837:8: warning: assignment makes pointer from integer without a cast
  value = strchrnul(line, ':');
        ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:870:2: warning: implicit declaration of function ‘stpcpy’ [-Wimplicit-function-declaration]
  t = stpcpy(t, line);
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:870:4: warning: assignment makes pointer from integer without a cast
  t = stpcpy(t, line);
    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_header_append_token’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:892:27: warning: assignment makes pointer from integer without a cast
  h->tokens[h->token_used] = strdup(token);
                           ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_header_serialize’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:922:4: warning: assignment makes pointer from integer without a cast
  t = stpcpy(t, h->key);
    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:925:4: warning: assignment makes pointer from integer without a cast
  t = stpcpy(t, h->value);
    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_serialize_common’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:1155:3: warning: implicit declaration of function ‘asprintf’ [-Wimplicit-function-declaration]
   r = asprintf(&head, "%s %s RTSP/%u.%u\r\n",
   ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:1196:6: warning: assignment makes pointer from integer without a cast
    p = stpcpy(p, m->body_headers[i].line);
      ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:1289:5: warning: assignment makes pointer from integer without a cast
   p = stpcpy(p, m->headers[i].line);
     ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:1302:4: warning: assignment makes pointer from integer without a cast
  p = stpcpy(p, head);
    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:1303:4: warning: assignment makes pointer from integer without a cast
  p = stpcpy(p, headers);
    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_message_from_request’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:1726:2: warning: implicit declaration of function ‘strncasecmp’ [-Wimplicit-function-declaration]
  if (strncasecmp(next, "RTSP/", 5))
  ^
In file included from /home/thopiekar/miraclecast/src/shared/shl_dlist.h:23:0,
                 from /home/thopiekar/miraclecast/src/shared/rtsp.c:35:
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_call_reply’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2594:6: note: in expansion of macro ‘rtsp_message_from_htable’
  m = rtsp_message_from_htable(elem);
      ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2594:6: note: in expansion of macro ‘rtsp_message_from_htable’
  m = rtsp_message_from_htable(elem);
      ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: note: each undeclared identifier is reported only once for each function it appears in
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2594:6: note: in expansion of macro ‘rtsp_message_from_htable’
  m = rtsp_message_from_htable(elem);
      ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_call’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2618:11: note: in expansion of macro ‘shl_dlist_entry’
   match = shl_dlist_entry(i, struct rtsp_match, list);
           ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2618:11: note: in expansion of macro ‘shl_dlist_entry’
   match = shl_dlist_entry(i, struct rtsp_match, list);
           ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2626:11: note: in expansion of macro ‘shl_dlist_entry’
   match = shl_dlist_entry(i, struct rtsp_match, list);
           ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_link_waiting’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:2676:11: error: ‘CLOCK_MONOTONIC’ undeclared (first use in this function)
           CLOCK_MONOTONIC,
           ^
In file included from /home/thopiekar/miraclecast/src/shared/shl_dlist.h:23:0,
                 from /home/thopiekar/miraclecast/src/shared/rtsp.c:35:
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_write’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:102:2: note: in expansion of macro ‘shl_dlist_entry’
  shl_dlist_entry(shl_dlist_first(head), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2825:6: note: in expansion of macro ‘shl_dlist_first_entry’
  m = shl_dlist_first_entry(&bus->outgoing, struct rtsp_message, list);
      ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:102:2: note: in expansion of macro ‘shl_dlist_entry’
  shl_dlist_entry(shl_dlist_first(head), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2825:6: note: in expansion of macro ‘shl_dlist_first_entry’
  m = shl_dlist_first_entry(&bus->outgoing, struct rtsp_message, list);
      ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_unref’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2951:8: note: in expansion of macro ‘shl_dlist_entry’
    m = shl_dlist_entry(i, struct rtsp_message, list);
        ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2951:8: note: in expansion of macro ‘shl_dlist_entry’
    m = shl_dlist_entry(i, struct rtsp_message, list);
        ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:98:20: note: in expansion of macro ‘rtsp_message_from_htable’
    _iter = _iter ? _accessor((void*)_iter) : NULL); \
                    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:181:2: note: in expansion of macro ‘SHL_HTABLE_FOREACH_MACRO’
  SHL_HTABLE_FOREACH_MACRO(_i, &(_bus)->waiting, rtsp_message_from_htable)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2959:4: note: in expansion of macro ‘RTSP_FOREACH_WAITING’
    RTSP_FOREACH_WAITING(m, bus) {
    ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:102:2: note: in expansion of macro ‘shl_dlist_entry’
  shl_dlist_entry(shl_dlist_first(head), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2969:9: note: in expansion of macro ‘shl_dlist_first_entry’
     m = shl_dlist_first_entry(&bus->outgoing,
         ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:108:15: note: in expansion of macro ‘rtsp_message_from_htable’
   htable__i ? _accessor(htable__i) : NULL; })
               ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:184:2: note: in expansion of macro ‘SHL_HTABLE_FIRST_MACRO’
  SHL_HTABLE_FIRST_MACRO(&(_bus)->waiting, rtsp_message_from_htable)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2975:16: note: in expansion of macro ‘RTSP_FIRST_WAITING’
    while ((m = RTSP_FIRST_WAITING(bus)))
                ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:102:2: note: in expansion of macro ‘shl_dlist_entry’
  shl_dlist_entry(shl_dlist_first(head), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:2987:11: note: in expansion of macro ‘shl_dlist_first_entry’
   match = shl_dlist_first_entry(&bus->matches,
           ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_attach_event’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:98:20: note: in expansion of macro ‘rtsp_message_from_htable’
    _iter = _iter ? _accessor((void*)_iter) : NULL); \
                    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:181:2: note: in expansion of macro ‘SHL_HTABLE_FOREACH_MACRO’
  SHL_HTABLE_FOREACH_MACRO(_i, &(_bus)->waiting, rtsp_message_from_htable)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3045:2: note: in expansion of macro ‘RTSP_FOREACH_WAITING’
  RTSP_FOREACH_WAITING(m, bus) {
  ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:98:20: note: in expansion of macro ‘rtsp_message_from_htable’
    _iter = _iter ? _accessor((void*)_iter) : NULL); \
                    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:181:2: note: in expansion of macro ‘SHL_HTABLE_FOREACH_MACRO’
  SHL_HTABLE_FOREACH_MACRO(_i, &(_bus)->waiting, rtsp_message_from_htable)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3045:2: note: in expansion of macro ‘RTSP_FOREACH_WAITING’
  RTSP_FOREACH_WAITING(m, bus) {
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3052:11: error: ‘CLOCK_MONOTONIC’ undeclared (first use in this function)
           CLOCK_MONOTONIC,
           ^
In file included from /home/thopiekar/miraclecast/src/shared/shl_dlist.h:23:0,
                 from /home/thopiekar/miraclecast/src/shared/rtsp.c:35:
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_detach_event’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:98:20: note: in expansion of macro ‘rtsp_message_from_htable’
    _iter = _iter ? _accessor((void*)_iter) : NULL); \
                    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:181:2: note: in expansion of macro ‘SHL_HTABLE_FOREACH_MACRO’
  SHL_HTABLE_FOREACH_MACRO(_i, &(_bus)->waiting, rtsp_message_from_htable)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3079:2: note: in expansion of macro ‘RTSP_FOREACH_WAITING’
  RTSP_FOREACH_WAITING(m, bus) {
  ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:98:20: note: in expansion of macro ‘rtsp_message_from_htable’
    _iter = _iter ? _accessor((void*)_iter) : NULL); \
                    ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:181:2: note: in expansion of macro ‘SHL_HTABLE_FOREACH_MACRO’
  SHL_HTABLE_FOREACH_MACRO(_i, &(_bus)->waiting, rtsp_message_from_htable)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3079:2: note: in expansion of macro ‘RTSP_FOREACH_WAITING’
  RTSP_FOREACH_WAITING(m, bus) {
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_remove_match’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3150:11: note: in expansion of macro ‘shl_dlist_entry’
   match = shl_dlist_entry(i, struct rtsp_match, list);
           ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_dlist.h:99:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((ptr), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3150:11: note: in expansion of macro ‘shl_dlist_entry’
   match = shl_dlist_entry(i, struct rtsp_match, list);
           ^
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_call_async’:
/home/thopiekar/miraclecast/src/shared/rtsp.c:3200:24: error: ‘CLOCK_MONOTONIC’ undeclared (first use in this function)
  m->timeout += shl_now(CLOCK_MONOTONIC);
                        ^
In file included from /home/thopiekar/miraclecast/src/shared/shl_dlist.h:23:0,
                 from /home/thopiekar/miraclecast/src/shared/rtsp.c:35:
/home/thopiekar/miraclecast/src/shared/rtsp.c: In function ‘rtsp_call_async_cancel’:
/home/thopiekar/miraclecast/src/shared/shl_macro.h:87:17: error: expected declaration specifiers or ‘...’ before ‘(’ token
   const typeof( ((_type *)0)->_member ) *__mptr = (_ptr); \
                 ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3249:6: note: in expansion of macro ‘rtsp_message_from_htable’
  m = rtsp_message_from_htable(elem);
      ^
/home/thopiekar/miraclecast/src/shared/shl_macro.h:88:22: error: ‘__mptr’ undeclared (first use in this function)
   (_type *)( (char *)__mptr - offsetof(_type, _member) ); \
                      ^
/home/thopiekar/miraclecast/src/shared/shl_htable.h:32:2: note: in expansion of macro ‘shl_container_of’
  shl_container_of((pointer), type, member)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:178:2: note: in expansion of macro ‘shl_htable_entry’
  shl_htable_entry((_p), struct rtsp_message, cookie)
  ^
/home/thopiekar/miraclecast/src/shared/rtsp.c:3249:6: note: in expansion of macro ‘rtsp_message_from_htable’
  m = rtsp_message_from_htable(elem);
      ^
src/shared/CMakeFiles/miracle-shared.dir/build.make:54: recipe for target 'src/shared/CMakeFiles/miracle-shared.dir/rtsp.c.o' failed
make[2]: *** [src/shared/CMakeFiles/miracle-shared.dir/rtsp.c.o] Error 1
CMakeFiles/Makefile2:136: recipe for target 'src/shared/CMakeFiles/miracle-shared.dir/all' failed
make[1]: *** [src/shared/CMakeFiles/miracle-shared.dir/all] Error 2
Makefile:117: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by albfan** [comment]
It's complicated to test miraclecast on a VM. The ethernet emulation does not support AP mode (as far as I know) I know because that was my first try when start with project.

About your problems I guess you are trying cmake config which is broken by now. See #7. autotools should work.

**Comment by thopiekar** [comment]
OK, I'll try witg autotools later.

What about connecting a USB stick to the VM? (Virtualbox)
Am 22.03.2015 11:34 schrieb "Alberto Fanjul" <notifications@github.com>:

**Comment by albfan** [comment]
Problem is on VM emulation. But if you solve it or get it to work would be a great advance.

NOTE: Do not include history when commenting from email (it add undesired noise)
NOTE2: Please stay close to actual issue. testing on VM is not related with testing on ubuntu. I have opened an issue for that #12

**Comment by albfan** [comment]
Ubuntu updates really quick. 

For a mere mortal (see me) is  unclear why a package comes with a test-suite that fails, or why ubuntu releases a package without commit it to it's official repo (now ubuntu systemd is on v219-4ubuntu8, but there's only tag until 219-4ubuntu5 see https://anonscm.debian.org/cgit/pkg-systemd/systemd.git/) I give up to understand.

Anyway all the fais seems to be fail tests, not fail compilation as

[src/test/test-udev.c:49](https://anonscm.debian.org/cgit/pkg-systemd/systemd.git/tree/src/test/test-udev.c?id=a10c00ed4ef365ad920286c047467fcae07c1d1b#n49) where path for rules.d is incorrect (ubuntu have it on /lib/udev/rules.d). After fix this test pass correctly

Another problem is related to [src/test/test-engine.c:95](https://anonscm.debian.org/cgit/pkg-systemd/systemd.git/tree/src/test/test-engine.c?id=a10c00ed4ef365ad920286c047467fcae07c1d1b#n95) where test8 is not passing (returns -22 EINVAL invalid argument)

I'm really thinking on close this issue and clone systemd, but ubuntu systemd changes so quickly that it will get obsolete almost instantly.

Surely `nocheck` could be safe to apply

**Comment by albfan** [comment]
it's not finished but patches on https://github.com/albfan/systemd-ubuntu-with-dbus allows to compile systemd with dbus on ubuntu

I'm dealing with 

    $ dpkg-source --commit

to generate patches on 

    $ apt-get source systemd

An compile succesfully a .deb package

**Comment by thopiekar** [comment]
Well, I think Ubuntu really has to fix all problems very soon, while looking at the release schedule: https://wiki.ubuntu.com/VividVervet/ReleaseSchedule
By the way, I managed to get miraclecast packaged, but when building it on launchpad it is still looking for cmake. Hope I'll get it fixed soon.

**Comment by albfan** [comment]
I managed to derive this issue to a new repo (it's not miraclecast issue to fix systemd ubuntu compilation)

https://github.com/albfan/systemd-ubuntu-with-dbus

Hope anyone can follow the guide on [README.md](https://github.com/albfan/systemd-ubuntu-with-dbus/blob/master/README.md) and get miraclecast working on ubuntu.

Please open issues related on that repo
